drivers: spi: spi_ll_stm32: Add SPI support for STM32MP1x SoC and stm32mp157c_dk2 board

yboutreux opened the pull request:
Abstract some SPI LL function call for future driver compatibility with a new SPI peripheral version (introduced with STM32MP1x and STM32H7x SoC).
Add SPI support for STM32MP1x SoC.
Add SPI support for stm32mp157c_dk2 board. If SPI is selected, SPI4 (Arduino connector compatible SPI) and SPI5 (on front 2x20 GPIO expander) will be enable by default on stm32mp157c_dk2 board.

More specific informations about "drivers: spi: spi_ll_stm32: set NSS before mode" commit : 
stm32mp1x (and stm32h7) LL function _SPI_Init_ seems to also define first the NSS then the mode, unlike other STM32 boards where this is not specified.
Changing the order shouldn't have bad repercussions on other boards, Zephyr SPI driver test has been passed successfully on disco_l475_iot1 board

yboutreux committed:
drivers: spi: spi_ll_stm32: set NSS before mode

This commit solves an issue where the NSS must be set before the
mode on the stm32mp157c_dk2, else LL_SPI_SetMode won't affect the
mode registry.
stm32mp1x (and stm32h7) LL function SPI_Init seems to also define
first the NSS then the mode, unlike other STM32 boards where this
is not specified.
Changing the order shouldn't have bad repercussions on other boards,
ZephyrnSPI driver test has been passed successfully on disco_l475_iot1
board.

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -318,6 +312,12 @@ static int spi_stm32_configure(struct device *dev,
 		}
 	}
 
+	if (config->operation & SPI_OP_MODE_SLAVE) {
```
If MP1 specific, are we sure this is safe for other series?
If yes, please state it in commit message.

yboutreux replied:
stm32mp1x (and stm32h7) LL function SPI_Init seems to also define first the NSS then the mode, unlike other STM32 boards where this is not specified.
Changing the order shouldn't have bad repercussions on other boards, Zephyr SPI driver test has been passed successfully on disco_l475_iot1 board

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -41,6 +48,8 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
 #define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | LL_SPI_SR_OVR)
 #endif
 
+#endif /* STM32MP1 */
```
CONFIG_SOC_SERIES_STM32MP1X

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -31,6 +31,13 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCE | LL_SPI_SR_MODF | \
+			   LL_SPI_SR_OVR | LL_SPI_SR_TIFRE)
+
+#else /* STM32MP1 */
```
CONFIG_SOC_SERIES_STM32MP1X

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -90,9 +99,26 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+	while (!LL_SPI_IsActiveFlag_TXP(spi)) {
+		/* NOP if not enough space for a packet*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
```
`LL_SPI_GetMode(spi)` likely doesn't return true/false, please test against `LL_SPI_MODE_MASTER`.
OTOH, I think `LL_SPI_IsEnabled` will logically return true here.

yboutreux replied:
`LL_SPI_IsEnabled` return either 1 (true) or 0 (false)
`LL_SPI_GetMode(spi)` return either :
```
#define LL_SPI_MODE_MASTER                      (SPI_CFG2_MASTER)
#define LL_SPI_MODE_SLAVE                          (0x00000000UL)
```

I agree this might not be super pretty to rely on the slave to be 0, and the master >0, do you prefer : 
`if (LL_SPI_IsEnabled(spi) && (LL_SPI_GetMode(spi) ==  LL_SPI_MODE_MASTER)) {`

erwango replied:
> LL_SPI_IsEnabled return either 1 (true) or 0 (false)

Well, at this step, should always be 1, isn't it?

yboutreux replied:
It should, I had odd behaviors and errors with the driver, and I had to check if the SPI is still enabled or if we encountered an issue that disabled the SPI.
Now this is fixed, and we don't need to check if the SPI is enabled, so this part of the condition has been removed

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -90,9 +99,26 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	/* STM32MP1 has some register name modifications :
+	 * LL_SPI_IsActiveFlag_TXP <-> LL_SPI_IsActiveFlag_TXE
```
In that case, might be good to add a level of abstraction here instead of ifdefing the whole code, cd what was done here:
https://github.com/zephyrproject-rtos/zephyr/tree/master/drivers/ipm/ipm_stm32_ipcc.c

yboutreux replied:
That's a good idea, I'll do that ! 
The equivalence is not straight forward, but I think the behavior should be the same  :

**Other MCUs : TXE**
0: Tx buffer not empty
1: Tx buffer empty
**MP1 : TXP**
0: there is not enough space to locate next data packet at TxFIFO
1: TxFIFO has enough free location to host 1 data packet
_________________________________________

**Other MCUs : RXNE**
0: Rx buffer empty
1: Rx buffer not empty
**MP1 : RXP**
0: RxFIFO is empty or a not complete data packet is received
1: RxFIFO contains at least 1 data packet

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -103,9 +129,18 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+	while (!LL_SPI_IsActiveFlag_RXP(spi)) {
+		/* NOP if buffer has less than a packet to read*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
```
ditto

yboutreux replied:
Same response :
`LL_SPI_IsEnabled` return either 1 (true) or 0 (false)
`LL_SPI_GetMode(spi)` return either :
```
#define LL_SPI_MODE_MASTER                      (SPI_CFG2_MASTER)
#define LL_SPI_MODE_SLAVE                          (0x00000000UL)
```

I agree this might not be super pretty to rely on the slave to be 0, and the master >0, do you prefer : 
`if (LL_SPI_IsEnabled(spi) && (LL_SPI_GetMode(spi) ==  LL_SPI_MODE_MASTER)) {`

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -178,22 +244,48 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
 	if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+		while (LL_SPI_IsEnabled(spi) &&
+		       !LL_SPI_IsActiveFlag_MODF(spi) &&
```
Not sure about these conditione, if we check what is done in the while block

yboutreux replied:
This is just an adaptation to the BSY flag which isn't existing on STM32MP1 SPI peripheral registers.
Definition taken from _stm32l4xx_ll_spi.h_ : 
```
/**
  * @brief  Get busy flag
  * @note   The BSY flag is cleared under any one of the following conditions:
  * -When the SPI is correctly disabled
  * -When a fault is detected in Master mode (MODF bit set to 1)
  * -In Master mode, when it finishes a data transmission and no new data is ready to be
  * sent
  * -In Slave mode, when the BSY flag is set to '0' for at least one SPI clock cycle between
  * each data transfer.
  * @rmtoll SR           BSY           LL_SPI_IsActiveFlag_BSY
  * @param  SPIx SPI Instance
  * @retval State of bit (1 or 0).
  */
```
Regarding the while loop content, this is something that is introduced here https://github.com/zephyrproject-rtos/zephyr/pull/17804

erwango commented on west.yml:
```diff
@@ -51,7 +51,7 @@ manifest:
       revision: 0ec40aed8087f26bd9ac1b70fb5a6c326a6451aa
       path: modules/hal/st
     - name: hal_stm32
-      revision: 686c93da4186eb9c48b83ee235fc0454693f6e5c
```
Split in dedicated commit

yboutreux replied:
Missed this comment, this is done thank you !

erwango commented on dts/arm/st/mp1/stm32mp157.dtsi:
```diff
@@ -147,6 +147,56 @@
 			};
 		};
 
+		spi1: spi@44004000 {
```
Split in dedicated commit, along with fixup changes

yboutreux replied:
Splitted, I also added changes related to STM32MP1x SoC, so commit are split in soc support, driver support and board support

erwango commented:
DNM for now (conflicts with STM32 I2C change in hal-stm32)

dbkinder reviewed (approved):
doc changes LGTM, thanks.

yboutreux commented:
Update : add abstraction to some LL function, since the adding of a new SPI peripheral overburden the code with some specific cases related to STM32MP1x SoC (and soon STM32H7x SoC which has the same SPI peripheral version)

yboutreux commented:
Rebased to remove merge conflicts

avisconti commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -307,12 +307,6 @@ static int spi_stm32_configure(struct device *dev,
 
 	LL_SPI_DisableCRC(spi);
 
```
"This commit solves an issue where the NSS must be set before the
mode on the stm32mp157c_dk2, else LL_SPI_SetMode won't affect the
mode registry."

Do you mean that there is an h/w issue you are taking care of?
Does it make sense to mention some offiicial documentation about it?

arnop2 replied:
@avisconti 
No it is a SW bug. The SPI configuration sequence is not respected (please refer to stm32MP1 ref manual for details). If NSS is configured after the mode is set then a MODF error can occur.

avisconti commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -25,57 +25,126 @@ struct spi_stm32_data {
+	LL_SPI_DisableIT_OVR(spi);
+	LL_SPI_DisableIT_CRCERR(spi);
+	LL_SPI_DisableIT_FRE(spi);
+	LL_SPI_DisableIT_MODF(spi);
```
I see all this changes for MP1. But is it really necessary to do it at zephyr level?
Shouldn't be a better idea to incorporate the SoC difference in LL driver?
In this case we will always call LL_SPI_DisableIT_ERR() which will do the proper thing for MP1.

erwango replied:
@avisconti, this is a separate topic that should be discussed internally.

erwango commented on soc/arm/st_stm32/stm32mp1/soc.h:
```diff
@@ -46,6 +46,10 @@
 #include <stm32mp1xx_ll_usart.h>
 #endif
 
+#ifdef CONFIG_SPI_STM32
```
This change should be with commit where you implement MP1 support in SPI driver

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -90,7 +90,7 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 
 	tx_frame = spi_stm32_next_tx(data);
-	while (!LL_SPI_IsActiveFlag_TXE(spi)) {
+	while (!spi_stm32_ll_func_IsActiveFlag_TXE(spi)) {
```
New functions, which are zephyr specific should follow zephyr coding style, hence should be lower_case (not CamelCase)

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -178,22 +179,22 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
-	LL_SPI_DisableIT_TXE(spi);
-	LL_SPI_DisableIT_RXNE(spi);
-	LL_SPI_DisableIT_ERR(spi);
+	spi_stm32_ll_func_DisableIT_TXE(spi);
```
Can we group these under `spi_stm32_ll_func_disable_it` ?

yboutreux replied:
Since we need to keep ll_func_enable_IT_TXE(spi), ll_func_enable_IT_RXNE(spi), ll_func_enable_IT_ERR(spi), ll_func_disable_IT_TXE(spi), ll_func_disable_IT_RXNE(spi), I thought that keeping ll_func_disable_IT_ERR(spi) instead of grouping those three lines was better for consistency sake

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -23,4 +23,59 @@ struct spi_stm32_data {
 	struct spi_context ctx;
 };
 
+static inline u32_t spi_stm32_ll_func_IsActiveFlag_TXE(SPI_TypeDef *spi)
```
Are these function names sufficiently abstracted?
Also I think you can spare the `spi_stm32` to minimize function names length.

yboutreux replied:
I see what you mean, we could take the opportunity to do "is_rx_empty" "is_tx_not_empty" "is_busy"... On my way !

yboutreux commented on boards/arm/stm32mp157c_dk2/Kconfig.defconfig:
```diff
@@ -24,4 +24,17 @@ config UART_7
+config SPI_4
+	default y
+
+config SPI_5
```
No you are right, is we enable SPI then all SPI should be available, thank you ! :)

yboutreux committed:
soc: arm: st_stm32: stm32mp1: Add SPI support

Add SPI support for STM32MP1x SoC.

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>

yboutreux committed:
boards: arm: stm32mp157c_dk2: Add SPI support

Add SPI support for stm32mp157c_dk2 board. If SPI is selected, SPI4
(Arduino connector compatible SPI) and SPI5 (on front 2x20 GPIO
expander) will be enable by default on stm32mp157c_dk2 board.

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -23,4 +23,59 @@ struct spi_stm32_data {
 	struct spi_context ctx;
 };
 
+static inline u32_t ll_func_TX_is_empty(SPI_TypeDef *spi)
```
Not fully coding style compliant yet...

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -31,6 +31,13 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
  */
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+
+#define LL_SPI_SR_CRCERR LL_SPI_SR_CRCE
```
Can we avoid this as well?
You're not supposed to (re)define LL code inside zephyr

yboutreux replied:
The goal was to "alias" the CRC error and TI frame error bits (which have the same purpose but a different name on this version of SPI peripheral)

We can also do this, but this chaining a bit more ifdefs : 

```
#ifdef CONFIG_SOC_SERIES_STM32MP1X

#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCE | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_TIFRE)

#else

#if defined(LL_SPI_SR_UDR)
#define SPI_STM32_ERR_MSK (LL_SPI_SR_UDR | LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
#elif defined(SPI_SR_FRE)
#define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | \
			   LL_SPI_SR_OVR | LL_SPI_SR_FRE)
#else
#define SPI_STM32_ERR_MSK (LL_SPI_SR_CRCERR | LL_SPI_SR_MODF | LL_SPI_SR_OVR)
#endif
#endif /* CONFIG_SOC_SERIES_STM32MP1X */
```

erwango replied:
@yboutreux this should be fine

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -25,57 +25,126 @@ struct spi_stm32_data {
 static inline void ll_func_set_FIFO_threshold_one_data(SPI_TypeDef *spi)
 {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	LL_SPI_SetFIFOThreshold(spi, LL_SPI_FIFO_TH_01DATA);
```
In case of other SPI FIFO IPs, threshold is set to 8-bit. Does 01data matches the same value?
In that case we can rename function as  `ll_func_set_fifo_threshold_8bit`

erwango reviewed (changes requested):
last mile to walk..

erwango commented:
Please rebase since dependency has been merged

yboutreux committed:
drivers: spi: spi_ll_stm32: Add support for STM32MP1x SoC

Add SPI driver support for STM32MP1x SoC.

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>

yboutreux commented:
> Please rebase since dependency has been merged

Change done and rebased, thank you !

yboutreux committed:
drivers: spi: st_stm32: Abstract LL function for spi ver. compatibility

Abstract some SPI LL function call for future driver compatibility with
a new SPI peripheral version (introduced with STM32MP1x and STM32H7x
SoC)

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>

erwango commented:
Approved, but need to merge https://github.com/zephyrproject-rtos/hal_stm32/pull/18 first.

yboutreux committed:
west.yml: update hal_stm32 revision to add SPI LL for CubeMP1

Update hal_stm32 revision in order to add SPI LL for CubeMP1

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
Signed-off-by: Alexandre Torgue <alexandre.torgue@st.com>
